src/JSInterop/Microsoft.JSInterop/src/DotNetObjectReferenceOfT.cs

```diff
@@ -50,10 +47,36 @@ internal long ObjectId
             get
             {
                 ThrowIfDisposed();
+                Debug.Assert(_objectId != 0, "Accessing ObjectId without tracking is always incorrect.");
+
                 return _objectId;
             }
         }
 
+        internal long TrackObjectReference(JSRuntime jsRuntime)
+        {
+            if (jsRuntime == null)
+            {
+                throw new ArgumentNullException(nameof(jsRuntime));
+            }
+
+            ThrowIfDisposed();
+
+            if (_jsRuntime is null)
+            {
+                _jsRuntime = jsRuntime;
+                _objectId = jsRuntime.TrackObjectReference(this);
+            }
+            else if (!ReferenceEquals(_jsRuntime, jsRuntime))
+            {
+                throw new InvalidOperationException($"{GetType().Name} is already being tracked by a different instance of {nameof(JSRuntime)}. A common cause is caching an instance of {nameof(DotNetObjectReference<TValue>)}" +
+                    $" globally. Consider creating instances of {nameof(DotNetObjectReference<TValue>)} at the JSInterop callsite instead");
+            }
+
```

Comment:
It's probably worth a comment here about why it's safe for a value to be tracked multiple times by the same runtime.